app/controllers/github_repos_controller.rb

```diff
@@ -5,77 +5,65 @@ class GithubReposController < ApplicationController
   def index
     authorize GithubRepo
 
-    client = create_octokit_client
-
-    existing_user_repos = current_user.github_repos.where(featured: true).
-      distinct.pluck(:github_id_code)
+    known_repositories_ids = current_user.github_repos.featured.distinct.pluck(:github_id_code)
 
-    @repos = client.repositories.map do |repo|
-      repo.selected = existing_user_repos.include?(repo.id)
-      repo
-    end
+    # NOTE: this will invoke autopaging, by issuing multiple calls to GitHub
+    # to fetch all of the user's repositories. This could eventually become slow
+    @repos = fetch_repositories_from_github(known_repositories_ids)
   rescue Octokit::Unauthorized => e
-    render json: { error: "Github Unauthorized: #{e.message}", status: 401 }, status: :unauthorized
-  end
-
-  def create
-    authorize GithubRepo
-    @repo = GithubRepo.find_or_create(fetched_repo_params(fetch_repo))
-    current_user.touch(:github_repos_updated_at)
-    if @repo.valid?
-      redirect_to "/settings/integrations", notice: "GitHub repo added"
-    else
-      redirect_to "/settings/integrations",
-                  error: "There was an error adding your Github repo"
-    end
-  end
-
-  def update
-    @repo = GithubRepo.find(params[:id])
-    current_user.touch(:github_repos_updated_at)
-    authorize @repo
-    if @repo.update(featured: false)
-      redirect_to "/settings/integrations", notice: "GitHub repo added"
-    else
-      redirect_to "/settings/integrations",
-                  error: "There was an error removing your Github repo"
-    end
+    render json: { error: "GitHub Unauthorized: #{e.message}", status: 401 }, status: :unauthorized
   end
 
   def update_or_create
     authorize GithubRepo
 
     params[:github_repo] = JSON.parse(params[:github_repo])
-    fetched_repo = fetch_repo
+
+    fetched_repo = fetch_repository_from_github(repo_params[:github_id_code])
     unless fetched_repo
-      render json: "error: Could not find Github repo", status: :not_found
+      render json: { error: "GitHub repository not found", status: 404 }, status: :not_found
       return
     end
 
-    repo = GithubRepo.find_or_create(fetched_repo_params(fetched_repo))
+    repo = GithubRepo.upsert(current_user, fetched_repo_params(fetched_repo))
 
     current_user.touch(:github_repos_updated_at)
 
     if repo.valid?
       render json: { featured: repo.featured }
     else
-      render json: "error: #{repo.errors.full_messages}"
+      render json: { error: repo.errors.full_messages, status: 422 }, status: :unprocessable_entity
     end
   end
 
   private
 
+  # TODO: use Github::UserClient or something
   def create_octokit_client
     current_user_token = current_user.identities.where(provider: "github").last.token
-    client = Octokit::Client.new(access_token: current_user_token)
-    client&.repositories&.sort_by!(&:name)
```

Comment:
this was a useless operation, as `client.repositories` invokes a HTTP call. Thus we were calling `.repositories` twice each time...